File: lib/core/utils/input_detector.dart
import 'package:lexi_core/features/dictionary/domain/entities/input_type.dart';

class InputDetector {
  InputDetector._();

  static const _phraseMaxWords = 4;
  static final _terminalPunctuation = RegExp(r'[.?!]\s*$');

  static InputType detect(String input) {
    final trimmed = input.trim();
    if (trimmed.isEmpty) return InputType.word;
    if (_terminalPunctuation.hasMatch(trimmed)) return InputType.sentence;

    final wordCount = trimmed.split(RegExp(r'\s+')).length;
    if (wordCount > _phraseMaxWords) return InputType.sentence;
    if (wordCount >= 2) return InputType.phrase;
    return InputType.word;
  }
}

File: apps/web/src/lib/inputDetector.ts
// Ports lib/core/utils/input_detector.dart's InputDetector.detect exactly —
// verified against that file, not guessed.
export type InputType = "word" | "phrase" | "sentence";

const PHRASE_MAX_WORDS = 4;
const TERMINAL_PUNCTUATION = /[.?!]\s*$/;

export function detectInputType(input: string): InputType {
  const trimmed = input.trim();
  if (trimmed === "") return "word";
  if (TERMINAL_PUNCTUATION.test(trimmed)) return "sentence";

  const wordCount = trimmed.split(/\s+/).length;
  if (wordCount > PHRASE_MAX_WORDS) return "sentence";
  if (wordCount >= 2) return "phrase";
  return "word";
}
